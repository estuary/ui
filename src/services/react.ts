import FullPageSpinner from 'components/fullPage/Spinner';
import { lazy } from 'react';
import { logRocketConsole, logRocketEvent } from './shared';
import { CustomEvents } from './types';

const LAZY_LOAD_FAILED_KEY = 'chunk_failed';

// https://mitchgavan.com/code-splitting-react-safely/
const setWithExpiry = (key: string, value: string, ttl: number) => {
    const item = {
        value,
        expiry: new Date().getTime() + ttl,
    };
    localStorage.setItem(key, JSON.stringify(item));
};

const getWithExpiry = (key: string) => {
    const itemString = window.localStorage.getItem(key);
    const item = itemString ? JSON.parse(itemString) : null;

    if (item === null) {
        return null;
    }

    // We have waited long enough to allow trying again so clearing out the key
    if (new Date().getTime() > item.expiry) {
        localStorage.removeItem(key);
        return null;
    }

    // Return value so we do not try reloading again
    return item.value;
};

const handledLazy = (factory: () => Promise<{ default: any }>) => {
    return lazy(() =>
        factory().catch((error) => {
            logRocketEvent(CustomEvents.LAZY_LOADING, 'failed');
            logRocketConsole('Component Failed Loading 1:', error);

            // Check if we're in an infinite loading loop before trying again
            if (!getWithExpiry(LAZY_LOAD_FAILED_KEY)) {
                setWithExpiry(LAZY_LOAD_FAILED_KEY, 'true', 10000);
                window.location.reload();
            }

            // Still make sure to return something so the app does not blow up
            return { default: FullPageSpinner };
        })
    );
};

export { handledLazy };
