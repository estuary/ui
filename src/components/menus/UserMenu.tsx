/* eslint-disable @typescript-eslint/no-unnecessary-condition */
import type { SxProps } from '@mui/material';

import { Stack, Typography } from '@mui/material';
import Divider from '@mui/material/Divider';
import ListItemIcon from '@mui/material/ListItemIcon';
import MenuItem from '@mui/material/MenuItem';

import { useShallow } from 'zustand/react/shallow';

import { LogOut, Mail, ProfileCircle } from 'iconoir-react';
import { FormattedMessage, useIntl } from 'react-intl';

import useEnhancedSupport from 'src/_compliance/hooks/useEnhancedSupport';
import IconMenu from 'src/components/menus/IconMenu';
import UserAvatar from 'src/components/shared/UserAvatar';
import { supabaseClient } from 'src/context/GlobalProviders';
import { useUserStore } from 'src/context/User/useUserContextStore';

interface Props {
    iconColor: string;
}

const nonInteractiveMenuStyling: SxProps = {
    '&:hover': {
        cursor: 'revert',
    },
};

const UserMenu = ({ iconColor }: Props) => {
    const intl = useIntl();
    const userDetails = useUserStore(useShallow((state) => state.userDetails));

    const { setEnhancedSupport: toggleEnhancedSupport } = useEnhancedSupport();

    const handlers = {
        logout: async () => {
            await supabaseClient.auth.signOut();
        },
    };

    if (userDetails) {
        const { avatar, email, emailVerified, userName } = userDetails;
        return (
            <IconMenu
                ariaLabel={intl.formatMessage({ id: 'accountMenu.ariaLabel' })}
                icon={
                    <UserAvatar
                        userEmail={email}
                        userName={userName}
                        avatarUrl={avatar}
                    />
                }
                identifier="account-menu"
                tooltip={intl.formatMessage({ id: 'accountMenu.tooltip' })}
            >
                <MenuItem sx={nonInteractiveMenuStyling}>
                    <ListItemIcon>
                        <ProfileCircle style={{ color: iconColor }} />
                    </ListItemIcon>
                    {userName}
                </MenuItem>

                <MenuItem sx={nonInteractiveMenuStyling}>
                    <ListItemIcon>
                        <Mail style={{ color: iconColor }} />
                    </ListItemIcon>

                    <Stack spacing={0}>
                        <Typography>{email}</Typography>

                        {emailVerified ? (
                            <Typography variant="caption">
                                <FormattedMessage id="accountMenu.emailVerified" />
                            </Typography>
                        ) : null}
                    </Stack>
                </MenuItem>

                <Divider />

                <MenuItem
                    onClick={() => {
                        void handlers.logout();
                    }}
                >
                    <ListItemIcon>
                        <LogOut style={{ color: iconColor }} />
                    </ListItemIcon>

                    <FormattedMessage id="cta.logout" />
                </MenuItem>

                <MenuItem
                    onClick={() => {
                        toggleEnhancedSupport(true);
                    }}
                >
                    Enable LogRocket
                </MenuItem>
            </IconMenu>
        );
    } else {
        return null;
    }
};

export default UserMenu;
