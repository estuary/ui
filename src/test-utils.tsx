// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render, RenderOptions } from '@testing-library/react';
import { ReactElement, type FC } from 'react';
import { act } from 'react-dom/test-utils';
import { localStorageKey } from './auth';
import AppProviders from './context';

const AllTheProviders: FC = ({ children }) => {
    return <AppProviders>{children}</AppProviders>;
};

const customRender = (
    ui: ReactElement,
    options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: AllTheProviders, ...options });

const loginAsUser = (userName: string = 'fakeUserName') => {
    act(() => {
        window.localStorage.setItem(localStorageKey, userName);
    });
};

export * from '@testing-library/react';
export { customRender as render, loginAsUser };
