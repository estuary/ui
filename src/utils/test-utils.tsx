/* eslint-disable @typescript-eslint/unbound-method */
// https://testing-library.com/docs/react-testing-library/setup#custom-render

import { render as rtlRender, RenderOptions } from '@testing-library/react';
import { AuthContext } from 'context/Auth';
import AppContent from 'context/Content';
import AppRouter from 'context/Router';
import AppTheme from 'context/Theme';
import { ReactElement } from 'react';
import { auth, AuthDetails } from 'services/auth';

// TODO - this does not really do anything because we hardcode
//   the username down below on the MockAuthProvider. However,
//   this will be needed in the future so leaving the work so far
//   so we can leverage this when wanting to test the "on load auth"
//   functionality
const loginAsUser = (username: string) => {
    const mockAuthDetails: AuthDetails = {
        session: {
            account_id: '',
            expires_at: '',
            token: '',
        },
        user: {
            created_at: '',
            display_name: username,
            email: `${username}@${username}`,
            id: '',
            name: username,
            unique_name: username,
            updated_at: '',
        },
    };

    auth.saveSession(mockAuthDetails.session);
    auth.saveUser(mockAuthDetails.user);
};

const logoutUser = () => {
    auth.removeAuthDetails();
};

const goTo = (route?: string, name?: string) => {
    window.history.pushState(
        {},
        name ? name : 'Test page',
        route ? route : '/'
    );
};

const waitForLoading = async () => {
    return Promise.resolve();
};

// Mocking AuthContext out basically skips all the loading screen and boostrapping
//  stuff that happens on load. This should be fine for unit testing as it
//  can help speed up tests a bit and makes things easier to deal with. This
//  does mean that we need to eventually write a basic test that handles the
//  account bootstrapping.
interface MockAuthProps {
    children: JSX.Element;
    username?: string;
}
const MockAuthProvider = ({ children, username }: MockAuthProps) => {
    const value = {
        login: jest.fn(() => {
            if (username) {
                loginAsUser(username);
            }
            return Promise.resolve();
        }),
        logout: jest.fn(() => {
            logoutUser();
            return Promise.resolve();
        }),
        user: username,
    };

    return (
        <AuthContext.Provider value={value}>{children}</AuthContext.Provider>
    );
};

const customRender = async (
    ui: ReactElement,
    options: Omit<RenderOptions, 'wrapper'> & {
        route?: string;
        user?: string;
    }
) => {
    const { route, user } = options;

    goTo(route, 'Test Page');

    const view = rtlRender(
        <AppContent>
            <MockAuthProvider username={user}>
                <AppTheme>
                    <AppRouter>{ui}</AppRouter>
                </AppTheme>
            </MockAuthProvider>
        </AppContent>,
        {
            ...options,
        }
    );

    await waitForLoading();

    return view;
};

export * from '@testing-library/react';
export { goTo, loginAsUser, logoutUser, customRender, waitForLoading };
