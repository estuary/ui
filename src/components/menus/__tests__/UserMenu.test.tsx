import { render } from '@testing-library/react';
import { renderOps } from 'test/test-utils';
import UserMenu from '../UserMenu';

const renderComponent = () => {
    console.log('renderComponent');
    return render(<UserMenu iconColor="red" />, renderOps);
};

describe('UserMenu', () => {
    test('should render', async () => {
        const rendered = renderComponent();

        expect(rendered.baseElement).toMatchSnapshot();
    });
});
