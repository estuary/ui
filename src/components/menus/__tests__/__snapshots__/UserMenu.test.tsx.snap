// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserMenu > should render 1`] = `
<body>
  <div />
</body>
`;
